File: queues/php/queue.php
<?php

class Node
{
    public $value;
    public ?Node $next = null;

    /**
     * Node constructor.
     *
     * @param $value
     */
    public function __construct($value, ?Node $next = null)
    {
        $this->value = $value;
        $this->next = $next;
    }
}


class Queue
{
    protected int $length = 0;
    protected ?Node $first = null;
    protected ?Node $last = null;

    public function isEmpty(): bool
    {
        return !$this->length;
    }

    /**
     * Add a value on to the end of the queue.
     *
     * @param $value mixed
     * @return int Number of nodes in the queue
     */
    public function enqueue($value): int
    {
        $newNode = new Node($value);

        if ($this->length === 0) {
            $this->first = $newNode;
        } else {
            $this->last->next = $newNode;
        }

        $this->last = $newNode;
        return ++$this->length;
    }

    /**
     * Remove the value from the front of the queue and return it.
     *
     * @return mixed
     */
    public function dequeue()
    {
        if ($this->length === 0) {
            throw new OutOfBoundsException('No values in queue');
        }

        $node = $this->first;
        $this->first = $node->next;
        --$this->length;
        return $node->value;
    }

    /**
     * Get the value at the front of the queue without removing it.
     *
     * @return mixed
     */
    public function peek()
    {
        if ($this->length === 0) {
            throw new OutOfBoundsException('No values in queue');
        }

        return $this->first->value;
    }
}

function println($string)
{
    echo $string . PHP_EOL;
}

$myQueue = new Queue();
println($myQueue->isEmpty() ? 'true' : 'false');

$myQueue->enqueue('Joy');
$myQueue->enqueue('Matt');
$myQueue->enqueue('Pavel');
$myQueue->enqueue('Samir');

println($myQueue->isEmpty() ? 'true' : 'false');

println($myQueue->peek());

println($myQueue->dequeue());
println($myQueue->dequeue());
println($myQueue->dequeue());
println($myQueue->dequeue());

println($myQueue->isEmpty() ? 'true' : 'false');

File: structures-and-algorithms/queues/python/queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.length = 0
        self.first = None
        self.last = None

    def is_empty(self):
        """ Is the queue empty? """
        return self.length == 0

    def enqueue(self, value):
        """ Add a value to the end of the queue. """
        new_node = Node(value)

        if self.length == 0:
            self.first = new_node
        else:
            self.last.next = new_node

        self.last = new_node
        self.length += 1
        return self.length

    def peek(self):
        """ Get the value from the front of the queue without removing it. """
        if self.length == 0:
            raise IndexError

        return self.first.value

    def dequeue(self):
        """ Get the value from the front of the queue and remove it. """
        if self.length == 0:
            raise IndexError

        node = self.first
        self.first = node.next
        self.length -= 1
        return node.value


my_queue = Queue()
print(my_queue.is_empty())

my_queue.enqueue('Joy')
my_queue.enqueue('Matt')
my_queue.enqueue('Pavel')
my_queue.enqueue('Samir')

print(my_queue.is_empty())

print(my_queue.peek())

print(my_queue.dequeue())
print(my_queue.dequeue())
print(my_queue.dequeue())
print(my_queue.dequeue())

print(my_queue.is_empty())
